packages/@aws-cdk/aws-apigatewayv2/lib/api.ts

```diff
@@ -0,0 +1,264 @@
+import * as lambda from '@aws-cdk/aws-lambda';
+import * as cdk from '@aws-cdk/core';
+import * as apigatewayv2 from '../lib';
+import { CfnApiProps } from './apigatewayv2.generated';
+
+/**
+ * the HTTP API interface
+ */
+export interface IApi extends cdk.IResource {
+  /**
+   * The ID of this API Gateway HTTP Api.
+   * @attribute
+   */
+  readonly apiId: string;
+}
+/**
+ * the ApiBase interface
+ */
+export interface IApiBase extends IApi {
+  /**
+   * http url of this api
+   */
+  readonly url: string
+}
+
+/**
+ * HttpProxyApi interface
+ */
+export interface IHttpProxyApi extends IApi {}
+/**
+ * LambdaProxyApi interface
+ */
+export interface ILambdaProxyApi extends IApi { }
+
+abstract class ApiBase extends cdk.Resource implements IApi {
+  /**
+   * API ID
+   */
+  public abstract readonly apiId: string;
+  /**
+   * API URL
+   */
+  public abstract readonly url: string;
+}
```

Comment:
It's not clear to me why we need all of these additional interface and abstract classes. It looks to me that we can drop these, can't we?

Same comment in `route.ts` as well.

Reply:
`IApi` is the interface of `Api` which creates `CfnApi` and from the L1 perspective, `CfnApi` is used to create either `WebSocket APIs` or `HTTP APIs` with either Lambda proxy integration or Http proxy integration. By design I implemented the `IApi` from the `Api` class for `CfnApi` so we can at least create `WebSocket APIs` with that(we yet to have L2 for it though).  To explicitly create HTTP API with Lambda proxy integration or HTTP proxy integration, I extended the `IApi` to `ILambdaProxyApi` and `IHttpProxyApi`, which `LambdaProxyApi` and `HttpProxyApi` implement them respectively.

I am not sure if this is a good design, but it's targeting a straightforward approach to create HTTP APIs with different integrations easily.

What do you think?

Reply:
> IApi is the interface of Api which creates CfnApi and from the L1 perspective, CfnApi is used to create either WebSocket APIs or HTTP APIs with either Lambda proxy integration or Http proxy integration. 

It seems you are approaching this based on the L1s. The reason we build L2s is to hide away the complexity of the L1s and provide a much better experience for the CDK customer.
I would completely change that thinking and interfaces and classes with the CDK customer in mind, and not what L1s are available.

> By design I implemented the IApi from the Api class for CfnApi so we can at least create WebSocket APIs with that(we yet to have L2 for it though).

Sounds a bit premature. People can always create websocket APIs by directly using the L1s. Let's focus on building a good Http API experience first. 

> What do you think?

I believe all we need is an interface called `IHttpApi`. I'm not yet seeing the value of any of the others, including `ApiBase`, `IApiBase`, `IHttpProxyApi`, etc.

Reply:
Thanks @nija-at. 

I will refactor it and just implement `IhttpApi` and `HttpApi` first.